File: Sources/Client/Public/Sources/Entity/Player.ts
import { decodeMood } from "../Native/Entity/Player/PlayerMood";
import { deltaTime } from "../../../Main";
import Entity from "./Entity";

// @UseRenderer(RendererFlower)
export default class Player extends Entity {
    public angryT: number;
    public sadT: number;

    public wasEliminated: boolean;

    /**
     * Player is dev flower or not.
     */
    public isDev: boolean;

    constructor(
        id: number,

        x: number,
        y: number,
        
        angle: number,

        size: number,

        health: number,

        public mood: number,

        public name: string,
    ) {
        super(id, x, y, angle, size, health);

        this.angryT = 0;
        this.sadT = 0;

        this.wasEliminated = false;

        this.isDev = false;
    }

    override update() {
        super.update();

        if (this.isDead) {
            this.sadT = 1;
            this.angryT = 0;
        } else {
            const interpolationRate = deltaTime / 100;

            const [isAngry, isSad] = decodeMood(this.mood);

            this.angryT = Math.min(1, Math.max(0, this.angryT + (isAngry ? interpolationRate : -interpolationRate)));
            this.sadT = Math.min(1, Math.max(0, this.sadT + (!isAngry && isSad ? interpolationRate : -interpolationRate)));
        }
    }
}

File: Sources/Client/private/src/Entity/Player.zig
const std = @import("std");
const math = std.math;
const Entity = @import("Entity.zig").Entity;
const pmood = @import("PlayerMood.zig");
const PureRenderer = @import("Renderers/Renderer.zig");

pub const Super = Entity(@This());

pub const Renderer = @import("Renderers/Player/PlayerRenderingDispatcher.zig").PlayerRenderingDispatcher;

comptime { // Validate
    PureRenderer.validateEntityImplementation(@This());
}

/// Mood of player.
mood: pmood.MoodBitSet = .initEmpty(),

/// Name of player.
name: []const u8,

/// Time properties for face pupil.
angry_t: f32 = 0,
sad_t: f32 = 0,

/// Whether this player is already eliminated from server yet.
was_eliminated: bool = false,

/// Whether this player is dev flower.
is_developer: bool = false,

pub fn init(
    _: std.mem.Allocator,
    name: []const u8,
) @This() {
    return .{ .name = name };
}

pub fn deinit(self: *@This(), _: std.mem.Allocator, _: *Super) void {
    self.* = undefined;
}

pub fn update(self: *@This(), delta_time: f32, entity: *Super) void {
    if (entity.is_dead) {
        self.sad_t = 1;
        self.angry_t = 0;
    } else {
        const mood_mouth_speed = delta_time / 100;

        var is_angry, var is_sad = pmood.decodeMood(self.mood);

        if (entity.is_poisoned) {
            is_angry = false;
            is_sad = false;
        }

        self.angry_t = math.clamp(self.angry_t + (if (is_angry) mood_mouth_speed else -mood_mouth_speed), 0, 1);
        self.sad_t = math.clamp(self.sad_t + (if (!is_angry and is_sad) mood_mouth_speed else -mood_mouth_speed), 0, 1);
    }
}
